### DISCUSSION TITLE
Using credentials_fetcher in a on-premises docker environment

### DISCUSSION THREAD

**Discussion by robsoncloud** [aws/credentials-fetcher#128] [Q&A]
I have created an ubuntu 22.04 image using the docker file available here but when running the ./credentials-fetcherd  --aws_sm_secret_name fake-secret-name and using the grpc_cli test I got the following error:

![image](https://github.com/aws/credentials-fetcher/assets/26840035/a72b438a-d038-4054-98f5-30d8cdf8e2e7)

```
Method name not found
Failed to find method AddKerberosLease in proto files.
Method name not found
Method name not found
Failed to parse request.
```

**grpc_cli command:**
```
grpc_cli call unix:/var/credentials-fetcher/socket/credentials_fetcher.sock AddKerberosLease "credspec_contents: '{\"CmsPlugins\":[\"ActiveDirectory\"],\"DomainJoinConfig\":{\"Sid\":\"S-1-5-21-1452704211-2353752802-397960360\",\"MachineAccountName\":\"webapp-gmsa\",\"Guid\":\"31e66e22-9755-4982-a244-1e8b5007c67a\",\"DnsTreeName\":\"intranet.local\",\"DnsName\":\"intranet.local\",\"NetBiosName\":\"INTRANET\"},\"ActiveDirectoryConfig\":{\"GroupManagedServiceAccounts\":[{\"Name\":\"webapp-gmsa\",\"Scope\":\"intranet.local\"},{\"Name\":\"webapp-gmsa\",\"Scope\":\"INTRANET\"}],\"HostAccountConfig\":{\"PluginGUID\":\"{859E1386-BDB4-49E8-85C7-3070B13920E1}\",\"PortableCcgVersion\":\"1\",\"PluginInput\":{\"CredentialArn\":\"arn:aws:secretsmanager:aws-region:111122223333:secret:MySecret\"}}}}'"
```

Using gmsa api test:

![image](https://github.com/aws/credentials-fetcher/assets/26840035/c3d35841-5ddc-40d1-8d2e-c82c901dc8b7)

The scenario below is supported?
I need my container that runs in my local docker server to connect to my domain controller, retrieve the gMSA token make it available to be used by my app.

**Comment by jamolina** [comment]
Hey @robsoncloud . Thanks for reaching out. We will take a look at the issue and get back to you as soon as possible.

**Reply by smhmhmd** [reply]
@robsoncloud 
I see the issue in a similar ubuntu setup, a small code-change is required.
I will discuss with @jamolina and @as14692, one of us will post the fix.

**Comment by robsoncloud** [comment]
Hi @smhmhmd 

First, I'd like to thank you for the latest PR. It has helped me make some progress in my tests.

Second, I have a few questions and would appreciate it if you could shed some light on them.

1. I was unable to use grpc_cli. It keeps saying that the AddKerberosLease method name was not found. Am I missing something?

```
root@6302b511d580:~/grpc/build# grpc_cli call unix:/var/credentials-fetcher/socket/credentials_fetcher.sock AddKerberosLease "credspec_contents: '{\"CmsPlugins\":[\"ActiveDirectory\"],\"DomainJoinConfig\":{\"Sid\":\"S-1-5-21-1452704211-2353752802-397960360\",\"MachineAccountName\":\"webapp-gmsa\",\"Guid\":\"31e66e22-9755-4982-a244-1e8b5007c67a\",\"DnsTreeName\":\"intranet.local\",\"DnsName\":\"intranet.local\",\"NetBiosName\":\"intranet\"},\"ActiveDirectoryConfig\":{\"GroupManagedServiceAccounts\":[{\"Name\":\"WebApp01\",\"Scope\":\"intranet.local\"},{\"Name\":\"WebApp01\",\"Scope\":\"intranet\"}]}}'"
Method name not found
Failed to find method AddKerberosLease in proto files.
Method name not found
Method name not found
Failed to parse request.
Reflection request not implemented; is the ServerReflection service enabled?
```

2. Since my environment runs in a Docker container without access to S3/Secret Manager, I haven't found any options other than using the gmsa_test_client and its parameters --create_kerberos_tickets_non_domain_joined to retrieve the Kerberos ticket.

On the first attempt, I got the error ERROR: ldapsearch not found in the service's stdout, which I fixed by installing the ldap-utils package.

Then, I encountered another error in credentials-fetcher.log:
```
2024-07-15 21:26:30     ERROR: Cannot get gMSA krb ticket
```

However, the credentials-fetcher stdout indicated everything was fine:
```
Using default cache: /tmp/krb5cc_0
Using principal: webapp-gmsa-user@INTRANET.LOCAL
Authenticated to Kerberos v5
Using default cache: /tmp/krb5cc_0
Using principal: webapp-gmsa-user@INTRANET.LOCAL
Authenticated to Kerberos v5
```

To confirm, I ran klist and saw the ticket was in the /tmp folder:
```
root@6302b511d580:~/credentials-fetcher/build/api/tests# klist
Ticket cache: FILE:/tmp/krb5cc_0
Default principal: webapp-gmsa-user@INTRANET.LOCAL

Valid starting     Expires            Service principal
07/15/24 21:30:37  07/16/24 07:30:37  krbtgt/INTRANET.LOCAL@INTRANET.LOCAL
        renew until 07/16/24 21:30:33
```

To test if the ticket was valid, I created a new shared folder dedicated to the gMSA user in my test DC01, installed smbclient, and then connected using the Kerberos ticket:
```
export KRB5CCNAME=/tmp/krb5cc_0
smbclient //dc01.intranet.local/temp -k

```
![image](https://github.com/user-attachments/assets/226d0711-bffe-4938-b819-309cd5808032)

The connection worked, and now I see the following message every 10 minutes:
```
2024-07-15 22:21:42     INFO: renewal started
```

After the tests, I am trying to find ways to use this in production. For the credspec, I was considering using the environment variable CF_CRED_SPEC_FILE to pass the file path as a mount volume in my container. However, I am unsure how to provide the credentials for the gMSA and who will initiate the initial ticket retrieval.

What is your suggestion for providing the credspec.json and passing down the user/password/domain since I have no access to S3/Secret Manager?

How my container entrypoint should look like ? 

``` Pseudo
RUN ticket retrieval
CMD ["./credentials-fetcher/"]
```


Thanks

**Comment by robsoncloud** [comment]
Despite it says that the renew process had started, the ticket is not being automatically renewed and  had expired:

Log
![image](https://github.com/user-attachments/assets/fdab6fa7-7993-47cf-a135-c02ee9aee1ee)

Ticket expired
![image](https://github.com/user-attachments/assets/4386b4a6-2243-44a6-b04d-709ee0589c17)

Packet capture:
![image](https://github.com/user-attachments/assets/8e56c1b9-4fb9-4b81-8a30-64c7e24da590)

I will reach out offline to see what can be done.

**Comment by smhmhmd** [comment]
Hi @robsoncloud ,

You should be using tickets from /var/credentials-fetcher/krbdir/.
Can you run 'ls -al /var/credentials-fetcher/krbdir/' and paste it here.

**Reply by smhmhmd** [reply]
Hi @robsoncloud 

Added documentation here - https://github.com/aws/credentials-fetcher/pull/132

**Reply by smhmhmd** [reply]
https://github.com/aws/credentials-fetcher?tab=readme-ov-file#examples

**Reply by robsoncloud** [reply]
@smhmhmd Hi this is the output from ls -al

![image](https://github.com/user-attachments/assets/a48b62d7-1444-43fc-9763-efcbd8744971)

**Reply by smhmhmd** [reply]
@robsoncloud 
Each of those directories represent lease-ids that you created, there should be a krb5cc kerberos ticket file inside each of those directories. You should be able to run `klist <krb5cc>' and check.
Using grpc, you can also delete the lease-ids.

**Reply by robsoncloud** [reply]
@smhmhmd 
Don't know how these leases folders were created since I used the gmsa_test_client to bootstrap the solution.

As you can see below, there are no files under the krbdir folders/subfolders:

![image](https://github.com/user-attachments/assets/18faabcc-f992-412e-ac62-80738c7fdb50)

btw, checking your examples I noted that you use the s3/secret manager to grab the username and password of the user that has permission to retrieve the gMSA account. 

`  # credentials-fetcherd --aws_sm_secret_name aws/directoryservices/d-xxxxxx/gmsa // Substitute your secret name in AWS secrets manager`

My scenario is completely detached from the cloud, so, I am not able to provide a s3 bucket or secret manager. Do we have a workaround for that?
